cabal-validate/src/Main.hs

```diff
@@ -0,0 +1,430 @@
+main :: IO ()
+main = do
+  opts <- parseOpts
+  forM_ (steps opts) $ \step -> do
+    runStep opts step
+
+-- | Run a given `Step` with the given `Opts`.
+runStep :: Opts -> Step -> IO ()
+runStep opts step = do
+  let title = displayStep step
+  printHeader title
+  let action = case step of
+        PrintConfig -> printConfig opts
+        PrintToolVersions -> printToolVersions opts
+        Build -> build opts
+        Doctest -> doctest opts
+        LibTests -> libTests opts
+        LibSuite -> libSuite opts
+        LibSuiteExtras -> libSuiteExtras opts
+        CliSuite -> cliSuite opts
+        CliTests -> cliTests opts
+        SolverBenchmarksTests -> solverBenchmarksTests opts
+        SolverBenchmarksRun -> solverBenchmarksRun opts
+        TimeSummary -> timeSummary opts
+  withTiming opts title action
+  T.putStrLn ""
+
+-- | Compiler with version number like @ghc-9.6.6@.
+baseHc :: Opts -> FilePath
+baseHc opts = "ghc-" <> showVersion (compilerVersion $ compiler opts)
+
+-- | Base build directory for @cabal-validate@.
+baseBuildDir :: Opts -> FilePath
+baseBuildDir opts = "dist-newstyle-validate-" <> baseHc opts
+
+-- | Absolute path to the build directory for this architecture.
+--
+-- This is a path nested fairly deeply under `baseBuildDir`.
+buildDir :: Opts -> FilePath
+buildDir opts =
+  cwd opts
+    </> baseBuildDir opts
+    </> "build"
+    </> archPath opts
+    </> baseHc opts
+
+-- | @--num-threads@ argument for test suites.
+--
+-- This isn't always used because some test suites are finicky and only accept
+-- @-j@.
+jobsArgs :: Opts -> [String]
+jobsArgs opts = ["--num-threads", show $ jobs opts]
+
+-- | Default arguments for invoking @cabal@.
+cabalArgs :: Opts -> [String]
+cabalArgs opts =
+  [ "--jobs=" <> show (jobs opts)
+  , "--with-compiler=" <> compilerExecutable (compiler opts)
+  , "--builddir=" <> baseBuildDir opts
+  , "--project-file=" <> projectFile opts
+  ]
+
+-- | The `buildDir` for @cabal-testsuite-3@.
+cabalTestsuiteBuildDir :: Opts -> FilePath
+cabalTestsuiteBuildDir opts =
+  buildDir opts
+    </> "cabal-testsuite-3"
+
+-- | Arguments for @cabal build@.
+cabalNewBuildArgs :: Opts -> [String]
+cabalNewBuildArgs opts = "build" : cabalArgs opts
+
+-- | Arguments for @cabal list-bin@.
+--
+-- This is used to find the binaries for various test suites.
+cabalListBinArgs :: Opts -> [String]
+cabalListBinArgs opts = "list-bin" : cabalArgs opts
+
+-- | Get the binary for a given @cabal@ target by running @cabal list-bin@.
+cabalListBin :: Opts -> String -> IO FilePath
+cabalListBin opts target = do
+  let args = cabalListBinArgs opts ++ [target]
+  stdout <-
+    readProcessStdout_ $
+      proc (cabal opts) args
+
+  pure (T.unpack $ T.strip $ T.toStrict $ T.decodeUtf8 stdout)
+
+-- | Get the RTS arguments for invoking test suites.
+--
+-- These seem to only be used for some of the test suites, I'm not sure why.
+rtsArgs :: Opts -> [String]
+rtsArgs opts =
+  case archPath opts of
```

Comment:
Another stylistic wibble that you don’t have to address: I dislike the plentiful applications to `opts` everywhere. My suggestion would be to use NamedFieldPuns and destruction of `opts` in the argument position.

Reply:
Yes, it's definitely not one of Haskell's strong suits. I originally tried to use `OverloadedRecordDot`, but of course we run CI with versions of GHC that don't support it.

I also want this script to be [as accessible as possible](https://www.haskellforall.com/2015/09/how-to-make-your-haskell-code-more.html), and `NamedFieldPuns` make it less clear where variables are declared.

I do agree the verbosity is unideal, though.